drivers/usb/udc/udc_sam0.c

```diff
@@ -0,0 +1,1257 @@
+#else
+#define NVM_USB_PAD_TRANSN_POS 45
+#define NVM_USB_PAD_TRANSN_SIZE 5
+	pad_transn = (*((uint32_t *)(NVMCTRL_OTP4) +
+			(NVM_USB_PAD_TRANSN_POS / 32)) >>
+		      (NVM_USB_PAD_TRANSN_POS % 32)) &
+		     ((1 << NVM_USB_PAD_TRANSN_SIZE) - 1);
+
+	if (pad_transn == 0x1F) {
+		pad_transn = 5U;
+	}
+#endif
+
+	base->PADCAL.bit.TRANSN = pad_transn;
+
+#ifdef USB_FUSES_TRANSP_ADDR
+#define NVM_USB_PAD_TRANSP_POS 50
+#define NVM_USB_PAD_TRANSP_SIZE 5
+	pad_transp = *(uint32_t *)USB_FUSES_TRANSP_ADDR;
+#else
+	pad_transp = (*((uint32_t *)(NVMCTRL_OTP4) +
+			(NVM_USB_PAD_TRANSP_POS / 32)) >>
+		      (NVM_USB_PAD_TRANSP_POS % 32)) &
+		     ((1 << NVM_USB_PAD_TRANSP_SIZE) - 1);
+
+	if (pad_transp == 0x1F) {
+		pad_transp = 29U;
+	}
+#endif
+
+	base->PADCAL.bit.TRANSP = pad_transp;
+
+#ifdef USB_FUSES_TRIM_ADDR
+	pad_trim = *(uint32_t *)USB_FUSES_TRIM_ADDR;
+#else
+#define NVM_USB_PAD_TRIM_POS 55
+#define NVM_USB_PAD_TRIM_SIZE 3
+	pad_trim = (*((uint32_t *)(NVMCTRL_OTP4) +
+		      (NVM_USB_PAD_TRIM_POS / 32)) >>
+		    (NVM_USB_PAD_TRIM_POS % 32)) &
+		   ((1 << NVM_USB_PAD_TRIM_SIZE) - 1);
+
+	if (pad_trim == 0x7) {
+		pad_trim = 3U;
+	}
+#endif
+
+	base->PADCAL.bit.TRIM = pad_trim;
+}
+
+static uint8_t sam0_get_bd_size(const uint16_t mps)
+{
+	switch (mps) {
+	case 8:
+		return 0;
+	case 16:
+		return 1;
+	case 32:
+		return 2;
+	case 64:
+		return 3;
+	case 128:
+		return 4;
+	case 256:
+		return 5;
+	case 512:
+		return 6;
+	case 1023:
+		return 7;
+	default:
+		__ASSERT(true, "Wrong maximum packet size value");
+		return 0;
+	}
+}
+
+static struct sam0_ep_buffer_desc *sam0_get_ebd(const struct device *dev, const uint8_t ep)
+{
+	const struct udc_sam0_config *config = dev->config;
+
+	return &config->bdt[USB_EP_GET_IDX(ep)];
+}
+
+static UsbDeviceEndpoint *sam0_get_ep_reg(const struct device *dev, const uint8_t ep)
+{
+	const struct udc_sam0_config *config = dev->config;
+	UsbDevice *const base = config->base;
+
+	return &base->DeviceEndpoint[USB_EP_GET_IDX(ep)];
+}
+
+static int sam0_prep_out(const struct device *dev,
+			 struct net_buf *const buf, struct udc_ep_config *const ep_cfg)
+{
+	UsbDeviceEndpoint *const endpoint = sam0_get_ep_reg(dev, ep_cfg->addr);
+	struct sam0_ep_buffer_desc *const bd = sam0_get_ebd(dev, ep_cfg->addr);
+	const uint16_t size = MIN(16383U, net_buf_tailroom(buf));
+	unsigned int lock_key;
+
+	if (!endpoint->EPSTATUS.bit.BK0RDY) {
+		LOG_ERR("ep 0x%02x buffer is used by the controller", ep_cfg->addr);
+		return -EBUSY;
+	}
+
+	lock_key = irq_lock();
+	if (ep_cfg->addr != USB_CONTROL_EP_OUT) {
+		bd->bank0.addr = (uintptr_t)buf->data;
+		bd->bank0.byte_count = 0;
+
+		bd->bank0.multi_packet_size = size;
+		bd->bank0.size = sam0_get_bd_size(udc_mps_ep_size(ep_cfg));
+	}
+
+	endpoint->EPSTATUSCLR.bit.BK0RDY = 1;
+	irq_unlock(lock_key);
+
+	LOG_DBG("Prepare OUT ep 0x%02x size %u", ep_cfg->addr, size);
+
+	return 0;
+}
+
+static int samp0_prep_in(const struct device *dev,
```

Comment:
Why `samp0_`  and not `sam0_`?

Reply:
Thanks @henrikbrixandersen. Attention test, you have passed.